src/invariant/BucketListStateConsistency.cpp

```diff
@@ -0,0 +1,300 @@
+// Copyright 2025 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "invariant/BucketListStateConsistency.h"
+#include "bucket/LedgerCmp.h"
+#include "bucket/LiveBucket.h"
+#include "invariant/InvariantManager.h"
+#include "ledger/InMemorySorobanState.h"
+#include "ledger/LedgerTypeUtils.h"
+#include "main/Application.h"
+#include "util/LogSlowExecution.h"
+#include "util/XDRCereal.h"
+#include <fmt/format.h>
+
+namespace stellar
+{
+BucketListStateConsistency::BucketListStateConsistency() : Invariant(true)
+{
+}
+
+// This checks for consistency between the in-memory snapshot and live
+// BucketList, along with other important properties. We check these properties:
+// 1. Every live entry in the BL is reflected in the in-memory cache
+// 2. No entry exists in the cache, but not the BL
+// 3. Each live soroban entry also has a live TTL entry associated with it
```

Comment:
nit: We should mention somewhere that we also verify that TTLs in cache match TTLs in BL (currently that's done in step 3)